app/utils/communication/notifications.py

```diff
@@ -0,0 +1,361 @@
+import logging
+
+import firebase_admin
+from fastapi import BackgroundTasks
+from firebase_admin import credentials, messaging
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from app.core.config import Settings
+from app.cruds import cruds_notification
+from app.models import models_notification
+from app.schemas.schemas_notification import Message
+from app.utils.types.notification_types import Topic
+
+hyperion_error_logger = logging.getLogger("hyperion.error")
+
+
+"""
+TODO
+
+Enlever les tokens qui ne marchent plus : https://firebase.google.com/docs/cloud-messaging/manage-tokens
+Rendre tout cela asynchrone
+
+Faire des methodes internes et des méthodes plus publiques
+
+"""
+
+
+class NotificationManager:
+    """
+    Notification manager for Firebase.
+    This class should only be instantiated once.
+    """
+
+    # See https://firebase.google.com/docs/cloud-messaging/send-message?hl=fr for documentation and examples
+
+    def __init__(self, settings: Settings):
+        self.use_firebase = settings.USE_FIREBASE
+
+        if not self.use_firebase:
+            hyperion_error_logger.info("Firebase is configured to be disabled.")
+            return
+
+        try:
+            firebase_cred = credentials.Certificate("firebase.json")
+            firebase_admin.initialize_app(firebase_cred)
+        except Exception as error:
+            hyperion_error_logger.error(
+                f"Firebase is not configured correctly, disabling the notification manager. Please check a valid firebase.json file exist at the root of the project. {error}."
+            )
+            self.use_firebase = False
+
+    async def _manage_firebase_batch_response(
+        self, response: messaging.BatchResponse, tokens: list[str], db: AsyncSession
+    ):
+        """
+        Manage the response of a firebase notification. We need to assume that tokens that failed to be send are not valid anymore and delete them from the database.
+        """
+        if response.failure_count > 0:
+            responses = response.responses
+            failed_tokens = []
+            for idx, resp in enumerate(responses):
+                if not resp.success:
+                    # The order of responses corresponds to the order of the registration tokens.
+                    failed_tokens.append(tokens[idx])
+            hyperion_error_logger.info(
+                f"{response.failure_count} messages failed to be send, removing their tokens from the database."
+            )
+            await cruds_notification.batch_delete_firebase_device_by_token(
+                tokens=tokens, db=db
+            )
+
+    async def _send_firebase_push_notification_by_tokens(
+        self,
+        db: AsyncSession,
+        data: dict[str, str] | None = None,
+        tokens: list[str] = [],
+    ):
+        """
+        Send a firebase push notification to a list of tokens.
+
+        We don't use this function in Hyperion, since we only use "trigger" notifications.
```

Comment:
Thus, can't we just delete it ?

Reply:
It is used by `_send_firebase_trigger_notification_by_tokens`

Reply:
Thus, the comment is not appropriate

Reply:
The two functions are used internally by the Notification Manager to send Trigger Notifications which are Notification without data but should not be used by Hyperion developers.

I changed the docstring to encourage the use of `_send_firebase_trigger_notification` instead.
Would you be okay with the wording of [d0abc2d](https://github.com/aeecleclair/Hyperion/pull/178/commits/d0abc2d0210353f756295936d2fd26760790eb3c)?

Reply:
Perfect !